draft of multithreading blog post (#408)
JuliaLang/www.julialang.org: blog/_posts/2019-07-14-multithreading.md

---
layout: post
title:  Announcing composable multi-threaded parallelism in Julia
author: Jeff Bezanson, Jameson Nash
---

Software performance depends more and more on exploiting multiple processor cores.
The [free lunch][] is still over.
Well, we here in the Julia developer community have something of a reputation for
caring about performance, so we've known for years that we would need a good
story for multi-threaded, multi-core execution.
Today we are happy to announce a major new chapter in that story.
We are releasing an entirely new threading interface for Julia programs:
fully general task parallelism, inspired by parallel programming systems

> like [Cilk][] and [Go][].

Reviewer: We should note in which version of Julia (commit or nightly) the new interface is available. Otherwise people may expect it in the latest release version.

Author: That's easy --- it's not available :)

In this paradigm, any piece of a program can be marked for execution in parallel,
and a "task" will be started to run that code automatically on an available thread.
A dynamic scheduler handles all the decisions and details for you.
Here's an example of parallel code you can now write in Julia:

```
function fib(n::Int)
    if n < 2
        return n
    end
    t = @par fib(n - 2)
    return fib(n - 1) + fetch(t)
end
```

> 

Reviewer: Pasting in the latest master exactly as is gives `@par not defined`.

This, of course, is the classic highly-inefficient tree recursive implementation of
the Fibonacci sequence --- but running on any number of processor cores!
The line `t = @par fib(n - 2)` starts a task to compute `fib(n - 2)`, which runs in
parallel with the following line computing `fib(n - 1)`.
`fetch(t)` waits for task `t` to complete and gets its return value.

This model of parallelism has many wonderful properties.
I think of it as somewhat analogous to garbage collection: with GC, you
can freely allocate objects without worrying about how it works or when and how they
are freed.

> With task parallelism, you freely spawn tasks without worrying about where they run.

Reviewer: It would be cool if we could say that a very large number of tasks (is it tens of thousands, or millions?) can be spawned without worry. Some users coming from HPC may have a pthreads view of the world, and this can help make it clear.

The model is portable and free from low-level details.
You don't need to explicitly start and stop threads, and you don't even need to know how
many processors or threads there are (though you can find out if you want).

The model is nestable and composable: you can start parallel tasks that call library
functions that themselves start parallel tasks, and everything works.
Your CPUs will not be over-subscribed with threads.
This property is crucial for a high-level language where a lot of work is done by library
functions.
You need to be free to write whatever code you need --- including parallel code ---
without worrying about how the libraries it calls are implemented.

This is, in fact, the reason we are excited about this announcement: from this point on,
multi-core parallelism is unleashed over the entire Julia package ecosystem.

## Some history

One of the most surprising aspects of this new feature is just how long it has been in
the works.
From the very beginning --- prior even to the 0.1 release --- Julia has had the `Task`
type, providing symmetric coroutines and event-based I/O.
So we have always had a unit of *concurrency* in the language, it just wasn't *parallel*

> (simultaneous streams of execution) yet.

Reviewer 2: Does this change to "real parallelism" here have a practical effect in some computation-based IO stuff, like managing a GPU and then doing CPU operations?

task scheduler [PARTR][] based on the idea of depth-first scheduling.
He sold all of us on it with some nice animated slides, and it also didn't hurt that
he was willing to do some of the work.
The plan was to first develop PARTR as a standalone C library so it could be tested
and benchmarked on its own, and then integrate it with the Julia runtime.

After Kiran completed the standalone version of PARTR, we embarked on a series of
work sessions including Anton Malakhov (also of Intel) to figure out how to do
the integration.
The Julia runtime brings many extra features, such as garbage collection and
event-based I/O, so this was not entirely straightforward.
Somewhat disappointingly, though not unusually for a complex software project,
it took much longer than expected --- nearly two years --- to get the new
system working reliably.
A later section of this post will explain some of the internals and difficulties
involved for the curious.
But first, let's take it for a spin.

## How to use it

To use Julia with multiple threads, set the `JULIA_NUM_THREADS` environment
variable:

```
$ JULIA_NUM_THREADS=4 ./julia

> ```

Reviewer 2: Note that IDEs like Juno automatically detect the number of cores in a user's processor, giving Julia multithreading out of the box when used in these systems.

Reviewer: Even if you automatically detect cores, you still need to be able to change the number - to give some cores to BLAS, or the OS or something else. So, ideally, Juno etc. would present the cores available, but have a way for you to change it in the IDE.

Reviewer 2: It does.

![Capture](https://user-images.githubusercontent.com/1814174/61190941-48d65780-a672-11e9-8bd6-25cfc2470eea.PNG)

state from another.
Therefore we decided to have the `seed!(n)` call affect only the current thread's
state.
That way, multiple independent code sequences that seed and then use random
numbers can at least individually work as expected.
For more elaborate seeding requirements, we recommend allocating and passing
your own RNG objects (e.g. `Rand.MersenneTwister()`).

## Under the hood

As with garbage collection, the simple interface (`@par`) belies great
complexity underneath.
Here I will try to summarize some of the main difficulties and design
decisions we faced.

### Allocating and switching task stacks

Each `Task` requires its own execution stack, distinct from the usual process or
thread stacks provided by Unix operating systems.
Windows has fibers, which correspond closely to tasks, and several library
implementations of similar abstractions exist for Unix-family systems.

There are many possible approaches to this, with different tradeoffs.
As we often do, we tried to pick a method that would maximize throughput
and reliability.
We have a shared pool of stacks allocated by `mmap` (`VirtualAlloc` on

> windows), defaulting to 4MiB each (2MiB on 32-bit systems).

Reviewer 2: Is this changeable?

recently-blocked task.
That works fine, but it means a task can exist in a strange intermediate state
where it is considered not to be running, and yet is in fact running the
scheduler.
In particular, we need to make sure no other thread sees that task and thinks
"oh, there's a task I can run", causing it to scribble on the scheduler's
stack.

### Classic bugs

While trying to get this new functionality working, we encountered several
maddeningly difficult bugs.
My hands-down favorite was a mysterious hang on Windows that was fixed
by literally [flipping a single bit][].

## Looking forward

While we are excited about this milestone, a lot of work remains.
Here are some of the points we hope to focus on to further develop
our threading capabilities:

* Performance work on task switch and I/O latency.
* Adding parallelism to the standard library. Many common operations like sorting and
  array broadcasting can now use multiple threads internally.
* Consider allowing task migration.
* Improved debugging tools.
* Explore API extensions like cancel points.
* Provide alternate schedulers.

## Acknowledgements

> We would like to gratefully acknowledge funding support from Intel and relational.ai

Reviewer: ```suggestion
We would like to gratefully acknowledge funding support from Intel and relationalAI
```

Reviewer 3: Are they ok with capitalizing it as "RelationalAI"?

---
layout: post
title:  Announcing composable multi-threaded parallelism in Julia
author: Jeff Bezanson, Jameson Nash
---

Software performance depends more and more on exploiting multiple processor cores.
The [free lunch][] is still over.
Well, we here in the Julia developer community have something of a reputation for

> caring about performance, so we've known for years that we would need a good

Reviewer 4: Although I sort of like the REUSE symbol for this:

![](https://previews.123rf.com/images/editkolase/editkolase1809/editkolase180900021/107435652-recycle-biodegradable-and-compostable-concept-or-reduce-reuse-recycle-concept-easy-to-modify.jpg)

           println("i = $i on thread $(Threads.threadid())")
       end
i = 1 on thread 1
i = 7 on thread 3
i = 2 on thread 1
i = 8 on thread 3
i = 3 on thread 1
i = 9 on thread 4
i = 10 on thread 4
i = 4 on thread 2
i = 5 on thread 2
i = 6 on thread 2
```

Without further ado, let's try some nested parallelism.
A perennial favorite example is mergesort, which divides its input in half
and recursively sorts each half.
The halves can be sorted independently, yielding a natural opportunity
for parallelism.
Here is the code:

```
# sort the elements of `v` in place, from indices `lo` to `hi` inclusive
function psort!(v, lo::Int=1, hi::Int=length(v))
    if lo >= hi                       # 1 or 0 elements; nothing to do
        return v
    end
    if hi - lo < 100000               # below some cutoff, run in serial
        sort!(view(v, lo:hi), alg = MergeSort)
        return v
    end

    mid = (lo+hi)>>>1                 # find the midpoint

    half = @par psort!(v, lo, mid)    # task to sort the lower half; will run
    psort!(v, mid+1, hi)              # in parallel with the current call sorting
                                      # the upper half

>     wait(half)                        # wait for the lower half to finish

Reviewer 4: I think it'd be fun to use `fetch` here (implementing `sort` instead of `sort!`), as that seems harder to me (relative to what other languages provide and do), and we're already making a copy below

Reviewer 4: ```julia
julia> function psort(v, lo::Int=1, hi::Int=length(v))
                   if lo >= hi                       # 1 or 0 elements; nothing to do
                       return v
                   end
                   if hi - lo < 100000               # below some cutoff, run in serial
                       return sort(view(v, lo:hi), alg = MergeSort)
                   end
               
                   mid = (lo+hi)>>>1                 # find the midpoint
               
                   half = @task psort(v, lo, mid)    # task to sort the lower half; will run
                   half.sticky = false
                   schedule(half)
                   right = psort(v, mid+1, hi)       # in parallel with the current call sorting
                                                     # the upper half
                   left = fetch(half)                # wait for the lower half to finish
                   out = similar(v, hi-lo+1)         # result
                   @assert length(right) + length(left) == length(out)
               
                   i, il, ir = 1, 1, 1               # merge the two sorted sub-arrays
                   @inbounds while il <= length(left) && ir <= length(right)
                       l, r = left[il], right[ir]
                       if l < r
                           out[i] = l
                           il += 1
                       else
                           out[i] = r
                           ir += 1
                       end
                       i += 1
                   end
                   @inbounds while il <= length(left)
                       out[i] = left[il]
                       il += 1
                       i += 1
                   end
                   @inbounds while ir <= length(right)
                       out[i] = right[ir]
                       ir += 1
                       i += 1
                   end
                   return out
               end
```
```
julia> using Random; Random.seed!(0); a = rand(20000000);

julia> @time sort(a);
  1.469319 seconds (6 allocations: 152.588 MiB)

julia> @time sort(a);
  1.540864 seconds (6 allocations: 152.588 MiB, 2.91% gc time)

julia> @time psort(a); # 1 thead
 20.526943 seconds (879.42 M allocations: 14.520 GiB, 9.55% gc time)

julia> @time psort(a); # 2 threads
 12.870170 seconds (879.42 M allocations: 14.520 GiB, 15.73% gc time)

julia> @time psort(a);
 10.782067 seconds (879.42 M allocations: 14.520 GiB, 18.62% gc time)

julia> @time psort(a); # 4 threads
  9.499449 seconds (879.42 M allocations: 14.520 GiB, 22.32% gc time)
```
```
      ┌────────────────────────────────────────┐ 
   30 │⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀│ 
      │⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀│ 
      │⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀│ 
      │⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀│ 
      │⢄⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀│ 
      │⠀⠉⠒⠤⡀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀│ 
      │⠀⠀⠀⠀⠈⠑⠢⢄⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀│ 
      │⠀⠀⠀⠀⠀⠀⠀⠀⠉⠒⢄⡀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀│ 
      │⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠈⠑⠦⠤⢄⣀⣀⣀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀│ 
      │⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠉⠉⠉⠒⠒⠒⠤⠤⠤⠤⢄⣀⣀⣀⣀⣀⠀⠀⠀⠀⠀│ 
      │⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠉⠉⠉⠉⠉│ 
      │⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀│ 
      │⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀│ 
      │⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀│ 
    0 │⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀│ 
      └────────────────────────────────────────┘ 
      1                                        4
```

---
layout: post
title:  Announcing composable multi-threaded parallelism in Julia
author: Jeff Bezanson, Jameson Nash
---

Software performance depends more and more on exploiting multiple processor cores.
The [free lunch][] from Moore's Law is still over.
Well, we here in the Julia developer community have something of a reputation for
caring about performance, so we've known for years that we would need a good
story for multi-threaded, multi-core execution.
Today we are happy to announce a major new chapter in that story.
We are releasing an entirely new threading interface for Julia programs:
general task parallelism, inspired by parallel programming systems

> like [Cilk][] and [Go][].

Reviewer 3: ```suggestion
like [Cilk][], [Intel Threading Building Blocks][] and [Go][].
```

could only appear in the runtime system outside any Julia frame --- a narrow window,
since of course we are usually executing Julia code.

## Looking forward

While we are excited about this milestone, a lot of work remains.
Here are some of the points we hope to focus on to further develop
our threading capabilities:

* Performance work on task switch and I/O latency.
* Adding parallelism to the standard library. Many common operations like sorting and
  array broadcasting could now use multiple threads internally.
* Consider allowing task migration.
* Improved debugging tools.
* Explore API extensions, e.g. cancel points.
* Provide alternate schedulers.
* Explore integration with the [TAPIR][] parallel IR (some early work [here][]).

## Acknowledgements

We would like to gratefully acknowledge funding support from Intel and relationalAI
that made it possible to develop these new capabilities.

We are also grateful to the several people who patiently tried this functionality
while it was in development and filed bug reports or pull requests, and spurred us
to keep going!

[free lunch]: http://www.gotw.ca/publications/concurrency-ddj.htm

> [Cilk]: http://cilk.mit.edu/

Reviewer 3: ```suggestion
[Cilk]: http://cilk.mit.edu/
[Intel Threading Building Blocks]: https://software.intel.com/en-us/intel-tbb/
```

We see it as somewhat analogous to garbage collection: with GC, you
freely allocate objects without worrying about when and how they are freed.
With task parallelism, you freely spawn tasks --- potentially millions of them --- without
worrying about where they run.

The model is portable and free from low-level details.
You don't need to explicitly start and stop threads, and you don't even need to know how
many processors or threads there are (though you can find out if you want).

The model is nestable and composable: you can start parallel tasks that call library
functions that themselves start parallel tasks, and everything works.
Your CPUs will not be over-subscribed with threads.
This property is crucial for a high-level language where a lot of work is done by library
functions.
You need to be free to write whatever code you need --- including parallel code ---
without worrying about how the libraries it calls are implemented.
(*in the future we plan to extend this to C libraries such as BLAS, currently it only applies to Julia code)

This is, in fact, the reason we are excited about this announcement: from this point on,

> multi-core parallelism is unleashed over the entire Julia package ecosystem.

Reviewer 5: ```suggestion
the possibility of adding multi-core parallelism is unleashed over the entire Julia package ecosystem.
```

maybe?

end
unlock(cond)
```

As in previous versions, the standard lock to use to protect critical sections
is `ReentrantLock`, which is now thread-safe (it was previously only used for
synchronizing tasks).
`Threads.SpinLock` is also available, to be used in rare circumstances where
(1) only threads and not tasks need to be synchronized, and (2) you expect to
hold the lock for a short time.
`Semaphore` and `Event` are also available, completing the standard set of
synchronization primitives.

### Thread-local state

Julia code naturally tends to be purely functional (no side effects or mutation),
or only uses local mutation, so migrating to full thread-safety will hopefully
be easy in many cases.
But if your code uses shared state and you'd like to make it thread-safe, there
is some work to do.
So far we have used two kinds of approaches to this in Julia's standard library:
synchronization (locks), and thread- or task-local state.
Locks work well for shared resources not accessed frequently, or for resources
that cannot be duplicated for each thread.

But for high-performance code we recommend thread-local state.
Our `psort!` routine above can be improved in this way.
Here is a recipe.

> First, we modify the function to accept pre-allocated buffers, using a default

Reviewer 4: ```suggestion
First, we modify the function signature to accept pre-allocated buffers, using a default
```

argument value to allocate space automatically when the caller doesn't provide it:

```

> function psort!(v, lo::Int=1, hi::Int=length(v), temps = [similar(v,cld(length(v),2)) for i = 1:Threads.nthreads()])

Reviewer 4: ```suggestion
function psort!(v, lo::Int=1, hi::Int=length(v), temps=[similar(v, cld(length(v), 2)) for i = 1:Threads.nthreads()])
```
```suggestion
function psort!(v, lo::Int=1, hi::Int=length(v), temps=[similar(v, (length(v) + 1) ÷ 2) for i = 1:Threads.nthreads()])
```
```suggestion
function psort!(v, lo::Int=1, hi::Int=length(v), temps=[similar(v, (hi - lo + 1) ÷ 2) for i = 1:Threads.nthreads()])
```
```suggestion
function psort!(v, lo::Int=1, hi::Int=length(v), temps=[similar(v, 0) for i = 1:Threads.nthreads()]) # and add from Base `(length(t) < m-lo+1) && resize!(t, m-lo+1)`
```

```

The maximum size of temporary array our mergesort needs is half the array, using
ceiling division (`cld`) to handle odd lengths.
We simply need to allocate one array per thread.
Next, we modify the recursive calls to reuse the space:

```
    half = @spawn psort!(v, lo, mid, temps)
    psort!(v, mid+1, hi, temps)
```

Finally, use the array reserved for the current thread, instead of allocating a new one:

```
    temp = temps[Threads.threadid()]
    copyto!(temp, 1, v, lo, m-lo+1)
```

After these minor modifications, let's check performance on our large
machine:

```
$ for n in 1 2 4 8 16; do    JULIA_NUM_THREADS=$n ./julia psort.jl; done
  2.723312 seconds (3.07 k allocations: 152.852 MiB, 0.14% gc time)
  1.711112 seconds (3.28 k allocations: 229.149 MiB, 0.59% gc time)
  0.971327 seconds (3.28 k allocations: 381.737 MiB, 1.60% gc time)
  0.782790 seconds (3.28 k allocations: 686.913 MiB, 8.63% gc time)
  0.722063 seconds (3.33 k allocations: 1.267 GiB, 21.43% gc time)
```

Definitely faster, but we do seem to have some work to do on the
scalability of the runtime system.

> ### Seeding the default random number generator

Reviewer 4: Move this after `### IO`?

Author: I put it here since I consider this something you might need to know to update code, while the IO section is more internal details.

scheduler.
In particular, we need to make sure no other thread sees that task and thinks
"oh, there's a task I can run", causing it to scribble on the scheduler's
stack.

### Classic bugs

While trying to get this new functionality working, we encountered several
maddeningly difficult bugs.
The clear favorite was a mysterious hang on Windows that was fixed
by literally [flipping a single bit][].

Another good one was a [missing exception handling personality][].
In hindsight that could have been straightforward, but was confounded by two
factors: first, the failure mode caused the kernel to stop our process in a way
that we were not able to intercept in a debugger, and second, the failure was
triggered by a seemingly-unrelated change.
All Julia stack frames have an exception handling personality set, so the problem
could only appear in the runtime system outside any Julia frame---a narrow window,
since of course we are usually executing Julia code.

## Looking forward

While we are excited about this milestone, a lot of work remains.
Here are some of the points we hope to focus on to further develop
our threading capabilities:

> * Performance work on task switch and I/O latency.

Reviewer 3: ```suggestion
We would like to gratefully acknowledge funding support from [Intel][] and [relationalAI][]
```

* Adding parallelism to the standard library. Many common operations like sorting and
  array broadcasting could now use multiple threads internally.
* Consider allowing task migration.
* Provide more atomic operations at the Julia level.
* Using multiple threads in the compiler.
* More performant parallel loops and reductions, with more scheduling options.
* Allow adding more threads at run time.
* Improved debugging tools.
* Explore API extensions, e.g. cancel points.
* Thread-safe data structures.
* Provide alternate schedulers.
* Explore integration with the [TAPIR][] parallel IR (some early work [here][]).

## Acknowledgements

> 

Reviewer 3: ```suggestion
[here]: https://github.com/JuliaLang/julia/pull/31086
[Intel]: https://www.intel.com/
[relationalAI]: http://relational.ai/
```

We would like to gratefully acknowledge funding support from Intel and relationalAI
that made it possible to develop these new capabilities.

We are also grateful to the several people who patiently tried this functionality
while it was in development and filed bug reports or pull requests, and spurred us

> to keep going!

Reviewer 4: ```suggestion
to keep going! We know there's remaining problems and so will appreciate you letting us know your experience with it through GitHub and Discourse channels!
```

This model of parallelism has many wonderful properties.
We see it as somewhat analogous to garbage collection: with GC, you
freely allocate objects without worrying about when and how they are freed.
With task parallelism, you freely spawn tasks---potentially millions of them---without
worrying about where they run.

The model is portable and free from low-level details.
You don't need to explicitly start and stop threads, and you don't even need to know how
many processors or threads there are (though you can find out if you want).

The model is nestable and composable: you can start parallel tasks that call library
functions that themselves start parallel tasks, and everything works.
Your CPUs will not be over-subscribed with threads.
This property is crucial for a high-level language where a lot of work is done by library
functions.
You need to be free to write whatever code you need---including parallel code---
without worrying about how the libraries it calls are implemented
(currently only for Julia code, but in the future we plan to extend this to native libraries
such as BLAS).

This is, in fact, the main reason we are excited about this announcement: from this point on,

> the possibility of adding multi-core parallelism is unleashed over the entire Julia package ecosystem.

Reviewer 5: ```suggestion
the capability of adding multi-core parallelism is unleashed over the entire Julia package ecosystem.
```

Re-thinking about this, I think this is a better word.

julia> b = copy(a); @time psort!(b);
  1.741141 seconds (3.78 k allocations: 686.935 MiB, 4.16% gc time)
```

While the run times are bit variable, we see a definite speedup from using
two threads.
The laptop we ran this on has four hyperthreads, and it is especially amazing
that the performance of this code continues to scale if we add a third thread:

```
julia> b = copy(a); @time psort!(b);
  1.511860 seconds (3.77 k allocations: 686.935 MiB, 6.45% gc time)
```

Thinking about this two-way decomposition algorithm running on three threads
can make your head hurt a little!
In our view, this helps underscore how "automatic" this interface makes
parallelism feel.

Let's try a different machine with more CPU cores:

```
$ for n in 1 2 4 8 16; do    JULIA_NUM_THREADS=$n ./julia psort.jl; done
  2.958881 seconds (3.58 k allocations: 686.932 MiB, 4.71% gc time)
  1.868720 seconds (3.77 k allocations: 686.935 MiB, 7.03% gc time)
  1.222777 seconds (3.78 k allocations: 686.935 MiB, 9.14% gc time)
  0.958517 seconds (3.79 k allocations: 686.935 MiB, 18.21% gc time)
  0.836891 seconds (3.78 k allocations: 686.935 MiB, 21.10% gc time)

> ```

Reviewer 4: can we graph this and/or show the normalized values?

still possible---more on that a little later.

## How to move to a parallel world

During the 1.3 series the new thread runtime is considered to be in beta testing.
An "official" version will appear in a later release, to give us time to settle
on an API we can commit to for the long term.
Here's what you need to know if you want to upgrade your code over this period.

### Task scheduling and synchronization

To aid compatibility, code will continue to run within a single thread by default.
When tasks are launched using existing primitives (`schedule`, `@async`), they
will run only within the thread that launches them.
Similarly, a `Condition` object (used to signal tasks when events occur) can only
be used by the thread that created it.
Attempts to wait for or notify conditions from other threads will raise errors.
Separate thread-safe condition variables have been added, and are available as
`Threads.Condition`.
This needs to be a separate type because thread-safe use of condition variables
requires acquiring a lock.
In Julia, the lock is bundled with the condition, so `lock` can simply be called
on the condition itself:

```
lock(cond::Threads.Condition)

> while !ready

Reviewer 4: ```suggestion
lock(cond::Threads.Condition)
try
    while !ready
        wait(cond)
    end
finally
    unlock(cond)
end
```

    wait(cond)
end
unlock(cond)
```

As in previous versions, the standard lock to use to protect critical sections
is `ReentrantLock`, which is now thread-safe (it was previously only used for
synchronizing tasks).

> `Threads.SpinLock` is also available, to be used in rare circumstances where

Reviewer 4: ```suggestion
There are some other types of locks defined internally for specific circumstances which usually should not be applicable to the typical user (these include `Threads.SpinLock`, `Threads.Mutex`, and a variety of libuv-based mutexes protecting various parts of the runtime). These are used in rare circumstances where (1) only threads and not tasks will be synchronized, and (2) you know the the lock will only be held for a short time.
```

> (1) only threads and not tasks need to be synchronized, and (2) you expect to

Reviewer 4: ```suggestion
```

hold the lock for a short time.

> `Semaphore` and `Event` are also available, completing the standard set of

Reviewer 4: not really complete? there'd also be barrier, rwlocks, and a "once" I think in a "standard set"
```suggestion
The `Threads` module also provides `Semaphore` and `Event` types with their standard definition.
```

> synchronization primitives.

Reviewer 4: ```suggestion
```
